### PR TITLE
[kube-prometheus-stack] fixes thanos configs

### PR DESCRIPTION
**Pull request by thameezb**
#### What this PR does / why we need it
Updates thanos configs to allow the usage of existing secrets or the option of creating a new secret object
Full design discussion can be found in the issue 

#### Which issue this PR fixes
fixes https://github.com/prometheus-community/helm-charts/issues/3869

#### Checklist
- [x] [DCO](https://github.com/prometheus-community/helm-charts/blob/main/CONTRIBUTING.md#sign-off-your-work) signed
- [x] Chart Version bumped
- [x] Title of the PR starts with chart name (e.g. `[prometheus-couchdb-exporter]`)

### CODE REVIEW

**Comment by GMartinez-Sisti** [comment]
Thanks @thameezb! Being a breaking change, we also need to include updating steps on the [README.md](https://github.com/prometheus-community/helm-charts/blob/main/charts/kube-prometheus-stack/README.md) file. Check this [example](https://github.com/prometheus-community/helm-charts/blob/main/charts/kube-prometheus-stack/README.md) to see what's expected.

**Comment by thameezb** [comment]
> Thanks @thameezb! Being a breaking change, we also need to include updating steps on the [README.md](https://github.com/prometheus-community/helm-charts/blob/main/charts/kube-prometheus-stack/README.md) file. Check this [example](https://github.com/prometheus-community/helm-charts/blob/main/charts/kube-prometheus-stack/README.md#from-41x-to-42x) to see what's expected.

Added!

File: charts/kube-prometheus-stack/values.yaml [hunk 1]
```diff
@@ -3961,16 +3961,24 @@ thanosRuler:
...
-    #     static_configs:
-    #       - alertmanager.thanos.io
-    #     scheme: http
-    #     timeout: 10s
+    alertmanagersConfig:
+      # use existing secret, if configured, alertmanagersConfig.secret will not be used
+      existingSecret: {}
+        # name: ""
+        # key: ""
+      # will render alertmanagersConfig
```
Comment (GMartinez-Sisti) [inline charts/kube-prometheus-stack/values.yaml:3969]: nit: make it more error proof (same for the other two
```suggestion
      # will render alertmanagersConfig secret data and configure it to be used by Thanos Ruler custom resource, ignored when alertmanagersConfig.existingSecret is set
```

File: charts/kube-prometheus-stack/values.yaml [hunk 2]
```diff
@@ -3987,7 +3995,21 @@ thanosRuler:
 
     ## ObjectStorageConfig configures object storage in Thanos. Alternative to
     ## ObjectStorageConfigFile, and lower order priority.
-    objectStorageConfig: {}
+    objectStorageConfig:
+      # use existing secret, if configured, objectStorageConfig.secret will not be used
+      existingSecret: {}
+        # name: ""
+        # key: ""
+      # will render objectStorageConfig secret data and configure it to be used by Thanos Ruler custom resource, ignored when alertmanagersConfig.existingSecret is set
```
Comment (GMartinez-Sisti) [inline charts/kube-prometheus-stack/values.yaml:4003]:
```suggestion
      # will render objectStorageConfig secret data and configure it to be used by Thanos Ruler custom resource, ignored when objectStorageConfig.existingSecret is set
```

**Comment by GMartinez-Sisti** [comment]
@QuentinBisson,other than the last suggestion above, for me this is good to go. Can you please take a look and let us know what you think? Also if we are not forgetting anything regarding the breaking change process. Thanks 🙏

**Comment by rouke-broersma** [comment]
@GMartinez-Sisti @QuentinBisson this does not include the same change for thanos sidecar. This makes the configuration of the `objectStorageConfig` inconsistent between different products.

**Comment by QuentinBisson** [comment]
@rouke-broersma this is a good finding yes :) @thameezb could you address this ?

**Comment by QuentinBisson** [comment]
@GMartinez-Sisti Other than the issue brought up by @rouke-broersma I'm fine with the rest :)

**Comment by thameezb** [comment]
> @GMartinez-Sisti @QuentinBisson this does not include the same change for thanos sidecar. This makes the configuration of the `objectStorageConfig` inconsistent between different products.
> 
> https://github.com/prometheus-community/helm-charts/blob/6fb47fbaba215588d054e92db5a9fd060a0fc9d1/charts/kube-prometheus-stack/values.yaml#L3434

@QuentinBisson added. Please verify (I am not 100% of my testing here as I have not used this aspect of the config before)

**Comment by dragoangel** [comment]
@thameezb I think we also need to remove `thanosRuler.thanosRulerSpec.objectStorageConfigFile` as it doesn't make sense before and even more - now. It just adds "confusion", but doesn't help in any way. What do you think? @GMartinez-Sisti @QuentinBisson

File: charts/kube-prometheus-stack/templates/prometheus/prometheus.yaml [hunk 3]
```diff
@@ -356,7 +356,16 @@ spec:
 {{- if not .Values.prometheus.agentMode }}
 {{- if .Values.prometheus.prometheusSpec.thanos }}
   thanos:
-{{ toYaml .Values.prometheus.prometheusSpec.thanos | indent 4 }}
+{{- omit .Values.prometheus.prometheusSpec.thanos "objectStorageConfig" | toYaml | indent 4 }}
```
Comment (QuentinBisson) [inline charts/kube-prometheus-stack/templates/prometheus/prometheus.yaml:359]: Why are you using omit ?

**Comment by QuentinBisson** [comment]
@rouke-broersma could you check that this is fine for the thanos sidecar?

File: charts/kube-prometheus-stack/templates/prometheus/prometheus.yaml (hunk 3 above)
Comment (thameezb) [reply charts/kube-prometheus-stack/templates/prometheus/prometheus.yaml:359]: - I remove `objectStorageConfig` key from being pulled in directly from the `.Values.prometheus.prometheusSpec.thanos` block (as our structure defined in `values.yaml` is not syntactically correct for the thanosconfig CRD
- Dump the rest of the block config as per previous behaviour
- `.Values.prometheus.prometheusSpec.thanos.objectStorageConfig` is set, then convert it from the values syntax to thanos CRD correct syntax

**Comment by rouke-broersma** [comment]
I'm getting an error on thanos sidecar with this config:

```
  prometheusSpec:
    thanos:
      objectStorageConfig:
        existingSecret:
          key: config-file.yaml
          name: thanos-secret
```

```
Error: YAML parse error on kube-prometheus-stack/templates/prometheus/prometheus.yaml: error converting YAML to JSON: yaml: line 91: did not find expected key
helm.go:84: [debug] error converting YAML to JSON: yaml: line 91: did not find expected key
```

**Comment by thameezb** [comment]
> I'm getting an error on thanos sidecar with this config:

```yaml
  prometheusSpec:
    thanos:
      objectStorageConfig:
        existingSecret:
          key: config-file.yaml
          name: thanos-secret
```

```
Error: YAML parse error on kube-prometheus-stack/templates/prometheus/prometheus.yaml: error converting YAML to JSON: yaml: line 91: did not find expected key
helm.go:84: [debug] error converting YAML to JSON: yaml: line 91: did not find expected key
``` 

fixed and tested for all 3 permutations

**Comment by thameezb** [comment]
> @thameezb I think we also need to remove `thanosRuler.thanosRulerSpec.objectStorageConfigFile` as it doesn't make sense before and even more - now. It just adds "confusion", but doesn't help in any way. What do you think? @GMartinez-Sisti @QuentinBisson

removed

**Comment by rouke-broersma** [comment]
I can confirm that the config now looks correct

**Review by GMartinez-Sisti — APPROVED** [review]

**Comment by GMartinez-Sisti** [comment]
Thank you everyone for the help! 🚀  I'll let it simmer for a couple of hours and do some local testing in the meantime. If no one finds anything we can merge it.

**Comment by dragoangel** [comment]
@thameezb @GMartinez-Sisti one thing was missed:
```yaml
  # Source: kube-prometheus-stack/templates/prometheus/prometheus.yaml
  apiVersion: monitoring.coreos.com/v1
  kind: Prometheus
  ...
    thanos:
-     objectStorageConfig:
-       key: objstore.yml
-       name: thanos-monitoring-config
```
as it used as:
```yaml
  thanos:
{{- with (omit .Values.prometheus.prometheusSpec.thanos "objectStorageConfig")}}
{{ toYaml . | indent 4 }}
{{- end }}
```

**Comment by dragoangel** [comment]
@thameezb @GMartinez-Sisti one thing was missed:
```yaml
  # Source: kube-prometheus-stack/templates/prometheus/prometheus.yaml
  apiVersion: monitoring.coreos.com/v1
  kind: Prometheus
  ...
    thanos:
-     objectStorageConfig:
-       key: objstore.yml
-       name: thanos-monitoring-config
```
here is a patch for a fix:
```diff
diff --git "a/charts/kube-prometheus-stack/templates/prometheus/prometheus.yaml" "b/charts/kube-prometheus-stack/templates/prometheus/prometheus.yaml"
index 238bc6e3..eaa4b900
--- "a/charts/kube-prometheus-stack/templates/prometheus/prometheus.yaml"
+++ "b/charts/kube-prometheus-stack/templates/prometheus/prometheus.yaml"
@@ -356,14 +356,11 @@ spec:
 {{- if not .Values.prometheus.agentMode }}
 {{- if .Values.prometheus.prometheusSpec.thanos }}
   thanos:
-{{- with (omit .Values.prometheus.prometheusSpec.thanos "objectStorageConfig")}}
-{{ toYaml . | indent 4 }}
-{{- end }}
-{{- if .Values.prometheus.prometheusSpec.thanos.objectStorageConfig.existingSecret }}
+{{- if .Values.thanosRuler.thanosRulerSpec.objectStorageConfig.existingSecret }}
     objectStorageConfig:
-      key: "{{.Values.prometheus.prometheusSpec.thanos.objectStorageConfig.existingSecret.key }}"
-      name: "{{.Values.prometheus.prometheusSpec.thanos.objectStorageConfig.existingSecret.name }}"
-{{- else if .Values.prometheus.prometheusSpec.thanos.objectStorageConfig.secret}}
+      key: "{{.Values.thanosRuler.thanosRulerSpec.objectStorageConfig.existingSecret.key }}"
+      name: "{{.Values.thanosRuler.thanosRulerSpec.objectStorageConfig.existingSecret.name }}"
+{{- else if .Values.thanosRuler.thanosRulerSpec.objectStorageConfig.secret }}
     objectStorageConfig:
       key: object-storage-configs.yaml
       name: {{ template "kube-prometheus-stack.fullname" . }}-prometheus

```

**Comment by dragoangel** [comment]
Also I think we don't need `charts\kube-prometheus-stack\templates\prometheus\secret.yaml` file, I don't see a point in having same data as for thanosRuler but in prometheus. It will just create blank secret at this point, as `.Values.prometheus.prometheusSpec.thanos.objectStorageConfig` wrong in same way as above ^^

**Comment by thameezb** [comment]
> @thameezb @GMartinez-Sisti also I think we don't need `charts\kube-prometheus-stack\templates\prometheus\secret.yaml` file, I don't see a point in having a separate secret with the same data as for thanosRuler but in prometheus. It will just create a blank secret at this point, as `.Values.prometheus.prometheusSpec.thanos.objectStorageConfig` wrong in the same way as above ^^

Not sure if I agree. The above patch supposes that one has thanosruler configured/deployed if one wishes to add thanos side car info to prometheus. Is this always true?

Could I be using the side car (therefore configuring `prometheus.prometheusSpec.thanos` without configuring `thanosRuler.thanosRulerSpec`? I think so

Additionally there is other config being passed in `.Values.prometheus.prometheusSpec.thanos` beyond `objectStorageConfig`

**Comment by dragoangel** [comment]
Another thing that kube-prometheus used not lucky version of Thanos Ruler (while Thanos Sidecar is up to date): https://github.com/thanos-io/thanos/releases/tag/v0.32.4 which fix critical issue with object storage

**Comment by rouke-broersma** [comment]
> Another thing that kube-prometheus used not lucky version of Thanos Ruler v0.32.3 (while Thanos Sidecar is up to date): https://github.com/thanos-io/thanos/releases/tag/v0.32.4 which fix critical issue with object storage, created PR: https://github.com/prometheus-community/helm-charts/pull/3935/files

Afaik the critical fixes relate to uploading blocks so thanos ruler shouldn't be affected.

**Comment by thameezb** [comment]
> Another thing that kube-prometheus used not lucky version of Thanos Ruler v0.32.3 (while Thanos Sidecar is up to date): https://github.com/thanos-io/thanos/releases/tag/v0.32.4 which fix critical issue with object storage, created PR: https://github.com/prometheus-community/helm-charts/pull/3935/files

Welcome to create another PR to address this (chart owners can speak to the process of an application update - as I'm not sure)

**Comment by dragoangel** [comment]
> > Another thing that kube-prometheus used not lucky version of Thanos Ruler v0.32.3 (while Thanos Sidecar is up to date): https://github.com/thanos-io/thanos/releases/tag/v0.32.4 which fix critical issue with object storage, created PR: https://github.com/prometheus-community/helm-charts/pull/3935/files
> 
> Welcome to create another PR to address this (chart owners can speak to the process of an application update - as I'm not sure)

Actually, Thanos Ruler is affected as it can "record" data, and when you use sharded Prometheus you will always use Thanos Ruler for all "recordings" and not Prometheus as in sharded setup Prometheus doesn't see "all view" and he doesn't look up data via Thanos Query as Thanos Ruler does. PR already merged anyway.

**Comment by dragoangel** [comment]
> > Another thing that kube-prometheus used not lucky version of Thanos Ruler v0.32.3 (while Thanos Sidecar is up to date): https://github.com/thanos-io/thanos/releases/tag/v0.32.4 which fix critical issue with object storage, created PR: https://github.com/prometheus-community/helm-charts/pull/3935/files
> 
> Afaik the critical fixes relate to uploading blocks so thanos ruler shouldn't be affected.

Actually, Thanos Ruler is affected as it can "record" data, and when you use sharded Prometheus you will always use Thanos Ruler for all "recordings" and not Prometheus, as in sharded setup Prometheus doesn't see "all view" and he doesn't look up for global data via Thanos Query as Thanos Ruler does. Most alerts also would be placed in Thanos Ruler, in my case only Thanos-related alerts are in Prometheus, so I get triggered right after updating to Thanos Ruler v0.32.3 as it fails to upload data. PR already merged anyway.

**Pull request merged**